lib/vrl/stdlib/src/parse_apache_log.rs

```diff
@@ -0,0 +1,298 @@
+
+        let value = arguments.required("value");
+        let format = arguments
+            .required_enum("format", &variants)?
+            .try_bytes()
+            .expect("format not bytes");
+
+        let timestamp_format = arguments.optional("timestamp_format");
+
+        Ok(Box::new(ParseApacheLogFn {
+            value,
+            format,
+            timestamp_format,
+        }))
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[
+            Example {
+                title: "parse apache common log",
+                source: r#"encode_json(parse_apache_log!(s'127.0.0.1 bob frank [10/Oct/2000:13:55:36 -0700] "GET /apache_pb.gif HTTP/1.0" 200 2326', "common"))"#,
+                result: Ok(
+                    indoc! {r#"s'{"host":"127.0.0.1","identity":"bob","message":"GET /apache_pb.gif HTTP/1.0","method":"GET","path":"/apache_pb.gif","protocol":"HTTP/1.0","size":2326,"status":200,"timestamp":"2000-10-10T20:55:36+00:00","user":"frank"}'"#},
+                ),
+            },
+            Example {
+                title: "parse apache combined log",
+                source: r#"encode_json(parse_apache_log!(s'127.0.0.1 bob frank [10/Oct/2000:13:55:36 -0700] "GET /apache_pb.gif HTTP/1.0" 200 2326 "http://www.seniorinfomediaries.com/vertical/channels/front-end/bandwidth" "Mozilla/5.0 (X11; Linux i686; rv:5.0) Gecko/1945-10-12 Firefox/37.0"', "combined"))"#,
+                result: Ok(
+                    indoc! {r#"s'{"agent":"Mozilla/5.0 (X11; Linux i686; rv:5.0) Gecko/1945-10-12 Firefox/37.0","host":"127.0.0.1","identity":"bob","message":"GET /apache_pb.gif HTTP/1.0","method":"GET","path":"/apache_pb.gif","protocol":"HTTP/1.0","referrer":"http://www.seniorinfomediaries.com/vertical/channels/front-end/bandwidth","size":2326,"status":200,"timestamp":"2000-10-10T20:55:36+00:00","user":"frank"}'"#},
+                ),
+            },
+            Example {
+                title: "parse apache error log",
+                source: r#"encode_json(parse_apache_log!(s'[01/Mar/2021:12:00:19 +0000] [ab:alert] [pid 4803:tid 3814] [client 147.159.108.175:24259] I will bypass the haptic COM bandwidth, that should matrix the CSS driver!', "error"))"#,
+                result: Ok(
+                    indoc! {r#"s'{"client":"147.159.108.175","message":"I will bypass the haptic COM bandwidth, that should matrix the CSS driver!","module":"ab","pid":4803,"port":24259,"severity":"alert","thread":"3814","timestamp":"2021-03-01T12:00:19+00:00"}'"#},
+                ),
+            },
+        ]
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ParseApacheLogFn {
+    value: Box<dyn Expression>,
+    format: Bytes,
+    timestamp_format: Option<Box<dyn Expression>>,
+}
+
+impl Expression for ParseApacheLogFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let bytes = self.value.resolve(ctx)?;
+        let message = bytes.try_bytes_utf8_lossy()?;
```

Comment:
Since this only results in `Err` when `bytes` is not `Value::Bytes`, it's safe to `unwrap`.

Reply:
Is there any downside to leaving it as `?`?

Reply:
Yes, `unwrap` will have no runtime cost in release builds, whereas `?` has at least one more branch.

Reply:
Two points to make here:

1. Let's not worry about one or two more branches, it really doesn't matter in the grand scheme of things w.r.t. Vector performance. I always like to point people to https://gist.github.com/jboner/2841832 when discussions happen around performance numbers that _just don't matter_.
2. In general, I would advise _against_ deciding that an `unwrap` is safe to use if you aren't 100% sure. In this case in particular, the invariant that a parameter with `kind: Bytes` means you can never receive anything else is no longer true because we moved back to progressive type checking (see #6535). This means two things:
   1. At _runtime_ a function needs to be ready to handle any type (which basically means just using `try_xxx()?`.
   2. At _compile-time_ a function doesn't need to mark itself as fallible if it received the wrong type, this is now handled by the compiler, so functions can keep the same type definition as before, pretending that they always receive the correct value kind.